### ISSUE TITLE
libGL error: No matching fbConfigs or visuals found, failed to load driver: swrast

### ISSUE THREAD

**Issue by AM4283** [Lojemiru/AM2R-Autopatcher-Linux#24]
Upon running the app image I get this:
![image](https://user-images.githubusercontent.com/78922336/113524373-36a1f880-957c-11eb-8574-ffcd0c97bab4.png)
The game does not launch whatsoever.
I am using Arch Linux on the i3 window manager, system specs are as shown:
![image](https://user-images.githubusercontent.com/78922336/113524395-651fd380-957c-11eb-8654-7bcc65880a6c.png)
Sadly, I cannot seem to play this game on my arch machine as the new launcher will not work for me either, due to another error. I have installed the required dependencies as detailed by the README. I have tried the app image on another window manager (OpenBox) to see if that was the issue, and the same error appeared. The game works fine on an install of Pop OS on my laptop, and I am at a loss as to what could be causing the issue - I am not a developer, and I am semi-new to Linux, so sorry in advance since I am not the greatest debugger.
Any help would be appreciated :)

**Comment by Lojemiru** [comment]
i3 user detected. Deploying snarky splash text... "Was a mouse really *that* expensive?"

Jokes aside, you're using an NVIDIA card on Linux. This will cause problems, but we can (probably) work around them. You'll need to have the **32-bit** version of your card's drivers installed, and need to use the `LD_PRELOAD` environment variable that points to the appropriate drivers installed on your device. Our Linux expert says it'll likely be `LD_PRELOAD=/usr/lib32/libGL.so`, but it could vary depending on your installation. You'll pass that environment variable to the game when launching it like this: `env "LD_PRELOAD=/usr/lib32/libGL.so" ./AM2R.AppImage`.

**Comment by AM4283** [comment]
> i3 user detected. Deploying snarky splash text... "Was a mouse really _that_ expensive?"
> 
> Jokes aside, you're using an NVIDIA card on Linux. This will cause problems, but we can (probably) work around them. You'll need to have the **32-bit** version of your card's drivers installed, and need to use the `LD_PRELOAD` environment variable that points to the appropriate drivers installed on your device. Our Linux expert says it'll likely be `LD_PRELOAD=/usr/lib32/libGL.so`, but it could vary depending on your installation. You'll pass that environment variable to the game when launching it like this: `env "LD_PRELOAD=/usr/lib32/libGL.so" ./AM2R.AppImage`.

I have the package `lib32-nvidia-utils` installed already, and setting the `LD_PRELOAD` variable to `/usr/lib32/libGL.so` produces another error:
![image](https://user-images.githubusercontent.com/78922336/113529215-e2554380-9590-11eb-92a8-ba5157b66571.png)
This is a list of nvidia packages I have installed:
![image](https://user-images.githubusercontent.com/78922336/113529469-8212d180-9591-11eb-886e-b033dfdc0bd6.png)
Trying to run the app image on another arch install (tried on xfce4 and i3), with an Nvidia GT 740 with the proprietary drivers was successful after setting the `LD_PRELOAD` variable, however I am not having any luck on my main machine.

UPDATE: Using the native version works perfectly by using `env "LD_PRELOAD=libcurl.so.3" ./AM2R` ! Its a shame that the app image is broken, but I will keep this issue up for the time being to see if there is a solution.

**Comment by Lojemiru** [comment]
Hmm, interesting that the AppImage is broken. @Miepee might have a better idea as to why.

**Comment by Miepee** [comment]
Heya @AM4283, can you try using running the AppImage like this instead?
`LD_PRELOAD=/usr/lib/libGL.so ./AM2R.AppImage`

Or, if that doesn't work, try `LD_PRELOAD=/usr/lib32/libGL.so ./AM2R.AppImage` instead. I personally never heard of the env command, nor of any of these warnings that are getting thrown there, and multiple people with Nvidia Cards have been able to run the AppImage with the above environment variable. That was tested on some pre 10xx cards, on a 1070, and a 20xx card as well.
It might very well just be, that 30xx cards don't work, I haven't had anyone be able to test those yet :/

**Comment by AM4283** [comment]
> Heya @AM4283, can you try using running the AppImage like this instead?
> `LD_PRELOAD=/usr/lib/libGL.so ./AM2R.AppImage`
> 
> Or, if that doesn't work, try `LD_PRELOAD=/usr/lib32/libGL.so ./AM2R.AppImage` instead. I personally never heard of the env command, nor of any of these warnings that are getting thrown there, and multiple people with Nvidia Cards have been able to run the AppImage with the above environment variable. That was tested on some pre 10xx cards, on a 1070, and a 20xx card as well.
> It might very well just be, that 30xx cards don't work, I haven't had anyone be able to test those yet :/

Running `LD_PRELOAD=/usr/lib/libGL.so ./AM2R.AppImage` still ends up with a libGL error


```
❯ LD_PRELOAD=/usr/lib/libGL.so ./AM2R.AppImage

***************************************
*     YoYo Games Linux Runner V1.3    *
***************************************	
CommandLine: -game game.unx
ExeName= /tmp/.mount_AM2R.AtipGzh/usr/bin/AM2R
MemoryManager allocated: 17920

***************************************
*     YoYo Games Runner v1.0(999)[r32908]    *
***************************************	
RunnerLoadGame: game.unx
RunnerLoadGame() - assets/game.unx
Checking if INIFile assets/options.ini Exists
!!!!No INI File assets/options.ini
Reading File assets/game.unx
Loaded File assets/game.unx(68872432)
IFF wad found
Get Resolution
Get Header Information
InitGMLFunctions
HighScore..SavePrePend /home/albi/.config/AM2R/
Game..Math..Graphic..Action..File..Resource..Interaction..3D..Particle..Misc..DS..Sound..Physics..Gamepad..GAMEPAD: Initialising Ubuntu support
GAMEPAD: 4 axis values (last) at 0xf6936660
GAMEPAD: 4 axis values (last) at 0xf6936640
GAMEPAD: 4 axis values (last) at 0xf6936620
GAMEPAD: 4 axis values (last) at 0xf6936600
Buffers..Networking..Shaders..InitPushFunctions...YoYo..Fini
ID_STRG
Create Window
Display Size(Pixels): 3840,1080
GameDisplayName=AM2R
Win #1
XF86VidModeExtension-Version 2.2
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Only Singlebuffered Visual!
glX-Version 1.4

```
and running `LD_PRELOAD=/usr/lib32/libGL.so ./AM2R.AppImage` ends up with a libGL error and "wrong ELF class" error
```
❯ LD_PRELOAD=/usr/lib32/libGL.so ./AM2R.AppImage
ERROR: ld.so: object '/usr/lib32/libGL.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/usr/lib32/libGL.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/usr/lib32/libGL.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/usr/lib32/libGL.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/usr/lib32/libGL.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/usr/lib32/libGL.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.

***************************************
*     YoYo Games Linux Runner V1.3    *
***************************************	
CommandLine: -game game.unx
ExeName= /tmp/.mount_AM2R.AyNYFeG/usr/bin/AM2R
MemoryManager allocated: 17920

***************************************
*     YoYo Games Runner v1.0(999)[r32908]    *
***************************************	
RunnerLoadGame: game.unx
RunnerLoadGame() - assets/game.unx
Checking if INIFile assets/options.ini Exists
!!!!No INI File assets/options.ini
Reading File assets/game.unx
Loaded File assets/game.unx(68872432)
IFF wad found
Get Resolution
Get Header Information
InitGMLFunctions
HighScore..SavePrePend /home/albi/.config/AM2R/
Game..Math..Graphic..Action..File..Resource..Interaction..3D..Particle..Misc..DS..Sound..Physics..Gamepad..GAMEPAD: Initialising Ubuntu support
GAMEPAD: 4 axis values (last) at 0xf6992660
GAMEPAD: 4 axis values (last) at 0xf6992640
GAMEPAD: 4 axis values (last) at 0xf6992620
GAMEPAD: 4 axis values (last) at 0xf6992600
Buffers..Networking..Shaders..InitPushFunctions...YoYo..Fini
ID_STRG
Create Window
Display Size(Pixels): 3840,1080
GameDisplayName=AM2R
Win #1
XF86VidModeExtension-Version 2.2
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Only Singlebuffered Visual!
glX-Version 1.4

```

**Comment by Miepee** [comment]
I deployed a (hopefully) working fix in the gpu-fix branch, if anyone here is still interested, testing it out would be appreciated

**Comment by Miepee** [comment]
Closed by new 2.2.0 AM2RLauncher release / merging of gpu-fix on Autopatcher-Linux.
Delete and reinstall the profile (go to the mod settings tab, open the `open profile content folder` button, delete the folder it opens).
Should you still have issues, feel free to reopen this.
